src/batEnum.ml

```diff
@@ -1237,6 +1237,27 @@ let merge test a b =
       n
     in append_from (append_from (from aux) a) b
 
+let interleave enums =
+  let enums_len = Array.length enums in
+  if not (enums_len > 0) then empty () else begin
+    let available = Array.make enums_len true
+    and next_idx = Array.init enums_len ((+) 1) in
+    next_idx.((Array.length next_idx) - 1) <- 0 ;
+    let rec next_elem idx =
+      match get enums.(idx) with
+       | Some x -> x , next_idx.(idx)
+       | None -> begin
+           available.(idx) <- false ;
+           let rec loop k =
+             let l = next_idx.(k) in
+             if l = idx then raise No_more_elements else
+             if available.(l) then (next_idx.(idx) <- l ; next_elem l) else loop l
+           in loop idx
+         end
+    in
+    from_loop 0 next_elem
+  end
+
 (*$T
```

Comment:
the test below is an inline test for the function `merge` above. You should not separate `merge` from its test (but put your implementation below the test), and also you should provide new tests for the function which did not exist before (see other `(*$T ... *)` blocks for inspiration). Inline tests can be tested with `make qtest` or `TESTABLE=src/batEnum.ml make qtest` to test only the `Enum` module.